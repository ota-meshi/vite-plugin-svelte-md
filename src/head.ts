import type { ResolvedOptions } from "./options"

export const headProperties = [
    "title",
    "meta",
    "link",
    "base",
    "style",
    "script",
    "htmlAttrs",
    "bodyAttrs",
]

export type HeadObject = Record<string, any>

/**
 * Preprocess head in frontmatter.
 */
export function preprocessHead(
    frontmatter: any,
    options: ResolvedOptions,
): HeadObject | null {
    if (!options.headEnabled) return frontmatter

    const head = frontmatter

    const meta = Array.isArray(head.meta) ? [...head.meta] : []

    if (head.title) {
        if (!meta.find((i: any) => i.property === "og:title"))
            meta.push({ property: "og:title", content: head.title })
    }

    if (head.description) {
        if (!meta.find((i: any) => i.property === "og:description"))
            meta.push({
                property: "og:description",
                content: head.description,
            })

        if (!meta.find((i: any) => i.name === "description"))
            meta.push({ name: "description", content: head.description })
    }

    if (head.image) {
        if (!meta.find((i: any) => i.property === "og:image"))
            meta.push({ property: "og:image", content: head.image })

        if (!meta.find((i: any) => i.property === "twitter:card"))
            meta.push({
                name: "twitter:card",
                content: "summary_large_image",
            })
    }

    const result: HeadObject = {}

    for (const [key, value] of Object.entries(head)) {
        if (headProperties.includes(key)) result[key] = value
    }
    if (meta.length > 0) {
        result.meta = meta
    }

    return Object.entries(result).length === 0 ? null : result
}

/**
 * Converts the header object to tags.
 */
export function headObjToTags(obj: HeadObject): string[] {
    const tags = []

    for (const key of Object.keys(obj)) {
        if (obj[key] == null) continue

        if (key === "title") {
            tags.push(`<title>${obj[key]}</title>`)
        } else if (key === "base") {
            tags.push(`<base ${attrs({ key: "default", ...obj[key] })}>`)
        } else if (headProperties.includes(key)) {
            const value = obj[key]
            if (Array.isArray(value)) {
                value.forEach((item) => {
                    tags.push(`<${key} ${attrs(item)}>`)
                })
            } else if (value) {
                tags.push(`<${key} ${attrs(value)}>`)
            }
        }
    }

    return tags

    /**
     * Convert to attributes string
     */
    function attrs(o: Record<string, string>) {
        return Object.entries(o)
            .map(([k, v]) => `${k}="${`${v}`.replace(/"/gu, "&quot;")}"`)
            .join(" ")
    }
}
