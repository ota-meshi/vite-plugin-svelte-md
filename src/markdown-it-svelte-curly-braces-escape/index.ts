import type MarkdownIt from "markdown-it"

/**
 * Escape curly braces in code block plugin
 */
export default function plugin(md: MarkdownIt): void {
    const originalCodeBlock = md.renderer.rules.code_block!
    md.renderer.rules.code_block = (...args) => {
        return escape(originalCodeBlock(...args))
    }
    const originalCodeInline = md.renderer.rules.code_inline!
    md.renderer.rules.code_inline = (...args) => {
        return escape(originalCodeInline(...args))
    }
    const originalFence = md.renderer.rules.fence!
    md.renderer.rules.fence = (...args) => {
        return escape(originalFence(...args))
    }

    /** Escape */
    function escape(test: string) {
        return test.replace(/[{}]/g, (c) => (c === "{" ? "&#123;" : "&#125;"))
    }
}
