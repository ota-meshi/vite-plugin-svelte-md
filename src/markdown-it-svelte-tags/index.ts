import type MarkdownIt from "markdown-it";

const SVELTE_TAGS_RE = /^(?:<svelte:[a-z][^>]*>|<\/svelte:[a-z]+>)/u;
/**
 * Svelte tags (e.g. `<svelte:head>`) plugin
 */
export default function plugin(md: MarkdownIt): void {
  md.inline.ruler.push("svelte-tags", (state) => {
    if (!state.md.options.html) {
      return false;
    }
    const pos: number = state.pos;

    const text: string = state.src.slice(pos);

    const match = SVELTE_TAGS_RE.exec(text);
    if (!match) {
      return false;
    }
    const token = state.push("html_inline", "", 0);
    token.content = state.src.slice(pos, pos + match[0].length);
    state.pos += match[0].length;
    return true;
  });
}
