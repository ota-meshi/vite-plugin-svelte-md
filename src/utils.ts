/** Normalize array */
export function toArray<T>(n: T | T[]): T[] {
    if (!Array.isArray(n)) return [n]
    return n
}
