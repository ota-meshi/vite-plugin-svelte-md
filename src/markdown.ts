import MarkdownIt from "markdown-it";
import grayMatter from "gray-matter";
import svelteTags from "./markdown-it-svelte-tags";
import svelteCurlyBracesEscape from "./markdown-it-svelte-curly-braces-escape";
import { toArray } from "./utils";
import { headObjToTags, preprocessHead } from "./head";
import type { ResolvedOptions } from "./options";

const SCRIPTS_RE = /(<script[^>]*>)([\s\S]*?)<\/script>/gu;
const SVELTE_TAGS_RE = /(<svelte:[a-z][^>]*>)([\s\S]*?)<\/svelte:[a-z]+>/gu;
const GET_SVELTE_TAG_NAME_RE = /^svelte:[a-z]+/u;
const IS_MODULE_CONTEXT_RE = /\bcontext\s*=\s*["']?module["']?/u;

const IS_SVELTE_TAG_NAME_RE = /^svelte:[a-z]+$/u;

class TagContent {
  private startTag = "";

  private readonly contents: string[] = [];

  public readonly tagName: string;

  private readonly defaultAttrs?: string;

  public constructor(tagName: string, defaultAttrs?: string) {
    this.tagName = tagName;
    this.defaultAttrs = defaultAttrs;
  }

  public toTag() {
    return `${
      this.startTag ||
      `<${this.tagName}${this.defaultAttrs ? ` ${this.defaultAttrs}` : ""}>`
    }
${this.contents.join("\n")}
</${this.tagName}>`;
  }

  public addTag(startTag: string, content: string) {
    if (!this.startTag) {
      this.startTag = startTag;
    }
    this.contents.push(content);
  }

  public add(content: string) {
    this.contents.push(content);
  }

  public prepend(content: string) {
    this.contents.unshift(content);
  }
}

/** Parse HTML */
function parseHtml(html: string) {
  const moduleContext = new TagContent("script", 'context="module"');
  const instanceScript = new TagContent("script");
  const svelteTags: TagContent[] = [];
  let newHtml = html.replace(SCRIPTS_RE, (_, startTag, script: string) => {
    let scriptContent = instanceScript;
    if (IS_MODULE_CONTEXT_RE.test(startTag)) {
      scriptContent = moduleContext;
    }
    scriptContent.addTag(startTag, script);
    return "";
  });
  newHtml = newHtml.replace(SVELTE_TAGS_RE, (_, startTag, inner) => {
    const tagName = GET_SVELTE_TAG_NAME_RE.exec(
      startTag.slice(1)
    )![0].toLowerCase();
    let svelteTag = svelteTags.find((tag) => tag.tagName === tagName);
    if (!svelteTag) {
      svelteTag = new TagContent(tagName);
      svelteTags.push(svelteTag);
    }
    svelteTag.addTag(startTag, inner);
    return "";
  });
  return { html: newHtml, moduleContext, instanceScript, svelteTags };
}

export type MarkdownProcessor = (id: string, text: string) => string;
/**
 * Creates md processor
 */
export function createMarkdownProcessor(
  options: ResolvedOptions
): MarkdownProcessor {
  const markdownIt = new MarkdownIt({
    html: true,
    linkify: true,
    typographer: true,
    ...options.markdownItOptions,
  });
  markdownIt.linkify.set({ fuzzyLink: false });
  // eslint-disable-next-line @typescript-eslint/unbound-method -- ignore
  const originalValidateLink = markdownIt.validateLink;
  markdownIt.validateLink = (url) => {
    if (!originalValidateLink(url)) {
      return false;
    }
    return !IS_SVELTE_TAG_NAME_RE.test(url);
  };
  markdownIt.use(svelteTags).use(svelteCurlyBracesEscape);

  options.markdownItUses.forEach((e) => {
    const [plugin, ...options] = toArray(e);

    markdownIt.use(plugin, ...options);
  });

  return (id: string, text: string) => {
    const raw = text.trimEnd();
    const { wrapperClasses, headEnabled } = options;

    const parsedFrontmatter = grayMatter(raw);

    const plainMarkdown = parsedFrontmatter?.content ?? raw;
    let html = markdownIt.render(plainMarkdown, { id });

    if (wrapperClasses) {
      html = `<div class="${wrapperClasses}">${html}</div>`;
    }

    const parsedHtml = parseHtml(html);

    const { head, frontmatter } = frontmatterPreprocess(
      parsedFrontmatter?.data ?? {},
      options
    );
    parsedHtml.moduleContext.prepend(
      `export const frontmatter = ${JSON.stringify(frontmatter)}`
    );
    if (headEnabled && head) {
      let svelteHead = parsedHtml.svelteTags.find(
        (tag) => tag.tagName === "svelte:head"
      );
      if (!svelteHead) {
        svelteHead = new TagContent("svelte:head");
        parsedHtml.svelteTags.push(svelteHead);
      }
      svelteHead.add(`${headObjToTags(head).join("\n")}`);
    }

    const svelteSfc = `${parsedHtml.moduleContext.toTag()}
${parsedHtml.instanceScript.toTag()}
${parsedHtml.svelteTags.map((tag) => tag.toTag()).join("\n")}
${parsedHtml.html}`;
    return svelteSfc;
  };
}

/** Process frontmatter */
function frontmatterPreprocess(frontmatter: any, options: ResolvedOptions) {
  const head = preprocessHead(frontmatter, options);
  return { head, frontmatter };
}
