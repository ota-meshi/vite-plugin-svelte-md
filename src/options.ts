import type MarkdownIt from "markdown-it";
import type { FilterPattern } from "@rollup/pluginutils";
import { toArray } from "./utils";

export interface Options {
  /**
   * Enable head support
   *
   * @default true
   */
  headEnabled?: boolean;

  /**
   * Options passed to Markdown It
   */
  markdownItOptions?: MarkdownIt.Options;

  /**
   * Plugins for Markdown It
   */
  markdownItUses?: (
    | MarkdownIt.PluginSimple
    | [MarkdownIt.PluginSimple | MarkdownIt.PluginWithOptions, any]
    | any
  )[];

  /**
   * Class names for wrapper div
   *
   * @default 'markdown-body'
   */
  wrapperClasses?: string | string[];

  include?: FilterPattern;
  exclude?: FilterPattern;
}

export interface ResolvedOptions extends Required<Options> {
  wrapperClasses: string;
}

/**
 * Resolve options
 */
export function resolveOptions(userOptions: Options): ResolvedOptions {
  const options: ResolvedOptions = {
    headEnabled: true,
    markdownItOptions: {},
    markdownItUses: [],
    include: null,
    exclude: null,
    ...userOptions,
    wrapperClasses: toArray(userOptions.wrapperClasses ?? "markdown-body")
      .filter((i) => i)
      .join(" "),
  };

  return options;
}
