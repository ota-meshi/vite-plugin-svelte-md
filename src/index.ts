import type { Plugin } from "vite";
import { createFilter } from "@rollup/pluginutils";
import { createMarkdownProcessor } from "./markdown";
import type { Options } from "./options";
import { resolveOptions } from "./options";

export type { Options };
/**
 * Creates vite-plugin-svelte-md
 */
export default function (options: Options = {}): Plugin {
  const resolvedOptions = resolveOptions(options);
  const mdToSvelte = createMarkdownProcessor(resolvedOptions);

  const filter = createFilter(
    resolvedOptions.include || /\.md$/,
    resolvedOptions.exclude
  );

  return {
    name: "vite-plugin-svelte-md",
    enforce: "pre",
    transform(raw, id) {
      if (!filter(id)) return undefined;
      try {
        return mdToSvelte(id, raw);
      } catch (e: any) {
        this.error(e);
      }
      return undefined;
    },
    handleHotUpdate(ctx) {
      if (!filter(ctx.file)) return;

      const defaultRead = ctx.read;
      ctx.read = async function () {
        return mdToSvelte(ctx.file, await defaultRead());
      };
    },
  };
}
