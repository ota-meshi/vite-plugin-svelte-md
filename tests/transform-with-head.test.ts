import { createMarkdownProcessor } from "../src/markdown"
import { resolveOptions } from "../src/options"
import chai from "chai"
import { jestSnapshotPlugin } from "mocha-chai-jest-snapshot"

chai.use(jestSnapshotPlugin())

describe("transform with head", () => {
    const options = resolveOptions({ headEnabled: true })
    const mdToSvelte = createMarkdownProcessor(options)

    it("basic", () => {
        const md = `---
title: Hey
---

# Hello

- A
- B
- C
`
        chai.expect(mdToSvelte("", md)).toMatchSnapshot()
    })

    it("head tag", () => {
        const md = `---
title: Hey
---

# Hello

- A
- B
- C

<script>
	import MyComponent from '$lib/MyComponent.svelte';
</script>

<svelte:head>
	<meta name="description" content="Bar" />
</svelte:head>

<MyComponent />
`
        chai.expect(mdToSvelte("", md)).toMatchSnapshot()
    })
})
