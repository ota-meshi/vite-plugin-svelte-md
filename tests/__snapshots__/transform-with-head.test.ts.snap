// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform with head basic 1`] = `
"<script context=\\"module\\">
export const frontmatter = {\\"title\\":\\"Hey\\"}
</script>
<script>

</script>
<svelte:head>
<title>Hey</title>
<meta property=\\"og:title\\" content=\\"Hey\\">
</svelte:head>
<div class=\\"markdown-body\\"><h1>Hello</h1>
<ul>
<li>A</li>
<li>B</li>
<li>C</li>
</ul>
</div>"
`;

exports[`transform with head head tag 1`] = `
"<script context=\\"module\\">
export const frontmatter = {\\"title\\":\\"Hey\\"}
</script>
<script>

	import MyComponent from '$lib/MyComponent.svelte';

</script>
<svelte:head>

<meta name=\\"description\\" content=\\"Bar\\" />

<title>Hey</title>
<meta property=\\"og:title\\" content=\\"Hey\\">
</svelte:head>
<div class=\\"markdown-body\\"><h1>Hello</h1>
<ul>
<li>A</li>
<li>B</li>
<li>C</li>
</ul>

<p></p>
<MyComponent /></div>"
`;
